[receiver/gitproviderreceiver] add branch time, commit, and line metrics

adrielp opened the pull request:
**Description:**

Adds the following branch based metrics.
* git.repository.branch.time
* git.repository.branch.commit.aheadby.count
* git.repository.branch.commit.behindby.count
* git.repository.branch.line.deletion.count
* git.repository.branch.line.addition.count
 
Additionally brings this receiver very close to the source version in the Liatrio Distro. Brings the testing up to par as well. After this pull request is merged, I'd like to move this component to alpha status so it can be built within the contrib binary and used by others. Once in alpha, I'd like to bring over the newest metric (repository.cve.count), bring over the GitLab scraper, and potentially propose a new component I developed to work in-tandem with this receiver, the GitHub App Authentication extension.

**Link to tracking Issue:**

 #22028

**Testing:** Brings testing inline with source and include scrape testing.

**Documentation:**

Updated the documentation to reflect the additional metrics.

adrielp committed:
[receiver/gitproviderreceiver] add branch age and commit metrics

adrielp committed:
[feat] add line addition and other metrics

adrielp committed:
[chore] gen changlog and update readme

Commit:
[chore] set go.mod in receiver to v1.21.0

Commit:
[chore] run make crosslink

Commit:
[chore] run make gotidy

crobert-1 commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -206,6 +210,92 @@ func (ghs *githubScraper) getPullRequests(
+	cc *string,
+	branchName string,
+) (*CommitNodeTargetCommitHistoryCommitHistoryConnection, error) {
+	data, err := getCommitData(context.Background(), client, repoName, ghs.cfg.GitHubOrg, 1, comCount, cc, branchName)
```
We might need to plumb down `ctx`, and keep a `cancel` function available for shutting down the scraper. Since the `genqlient` is making a HTTP request call using the given context, there should be some way to stop the request when the scraper is being shutdown.

(Let me know if I'm missing something here though)

adrielp replied:
Yup, that's a totally fair and solid point.

adrielp replied:
I passed in the `ctx` from the scraper. Let me know if there's something else to adjust here.

crobert-1 commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -206,6 +210,92 @@ func (ghs *githubScraper) getPullRequests(
+	}
+
+	if len(data.Repository.Refs.Nodes) == 0 {
+		return nil, errors.New("no commits returned")
```
Is there a valid use case where a branch exists but doesn't have commits? Should that be an error?

adrielp replied:
It's a possibility. I've updated the code/comments to reflect. Let me know your thoughts.

crobert-1 replied:
Makes sense, thanks for explaining 👍

crobert-1 commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -53,6 +53,41 @@ metrics:
+  git.repository.branch.time:
+    enabled: true
+    description: Time the branch has existed.
+    unit: "{branch}"
```
I don't think `branch` makes sense as the unit for these newly added metrics.  For this, would `s` make sense, similar to the `git.repository.pull_request.time_open` metric below?

(Same comment for the rest of the added metrics, their units should be updated as well)

adrielp replied:
Happy to adjust. We originally had them as `s` but adjusted to ` branch` based on some change in understanding on conventions. After re-reviewing the semconv on this, I agree it should be `s`. The previous change was a misunderstanding. Thanks for the catch.

adrielp replied:
Adjusted all metrics after re-reading the conventions.

crobert-1 reviewed (commented):
Looks good overall! Hopefully my comments were helpful, I don't have much experience with this receiver yet.

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -108,12 +108,45 @@ func (ghs *githubScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+					branch.Compare.BehindBy, branch.Name, branch.Compare.AheadBy, branch.Repository.Name)
+
+				// Yes, this looks weird. The aheadby metric is referring to the number of commits the branch is AHEAD OF the
+				// default branch, which in the context of the query is the behind by value. See the above below comment about
```
I'll update this comment, that's certainly a typo from the source.

adrielp replied:
This comment has been significantly updated.

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -206,6 +210,92 @@ func (ghs *githubScraper) getPullRequests(
+	branch BranchNode,
+) (int, int, int64, error) {
+	comCount := 100
+	var cc *string
```
I'll opt for `cursor`

adrielp replied:
I'll adjust. I'm about to refactor some of this due to some of your other questions. Mostly it'll be comments and renames.

adrielp replied:
updated to just cursor for clarity.

adrielp committed:
[chore] update metric types to match semantic conventions

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -206,6 +210,92 @@ func (ghs *githubScraper) getPullRequests(
+	// the number of commits made to the queried branch but not merged into
+	// the default branch. Doing it this way involves less queries because
+	// we don't have to know the queried branch name ahead of time.
+	comPages := getNumPages(float64(100), float64(branch.Compare.BehindBy))
```
They're mostly 100s, but there are some that are less than that. It refers to the number of items to be returned by a request. GitHub has a hard limit of 100, but in some cases less is preferable. Let me see what I can do.

adrielp replied:
Set a new constant called `defaultReturnItems` which is set to a 100. This takes the place of using 100 and in a couple of the sections where the default isn't used, I added comments as to why.

adrielp committed:
[chore] slight refactor with var and function renaming

includes query restructure and extensive comments for context on things
that had to be done based on github graphql query requirements

adrielp committed:
Merge branch 'main' into branch-metrics

Commit:
[chore] rerun tidy after merge conflicts

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -108,12 +108,45 @@ func (ghs *githubScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+			// metrics
+			for _, branch := range branches {
+				// Check if the branch is the default branch or if it is not behind the default branch
+				if branch.Name == branch.Repository.DefaultBranchRef.Name || branch.Compare.BehindBy == 0 {
```
@crobert-1 I'm still running through a few other things, but this comment has been updated to reflect the reasoning here. 

I'm thinking that it might be good for me to add a `limitations` section within the docs for receiver under the GitHub section. These limitations are different for each vendor. For example, when we were developing the GitLab scraper for this receiver, we weren't able to populate the Approval Time metric because it's not exposed via either the REST nor the GraphQL APIs. Thoughts here?

adrielp replied:
Added a limitations section.

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -108,12 +108,63 @@ func (ghs *githubScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+				var deletions int
+				var age int64
+
+				// TODO: rename getCommitInfo
```
Reminder for me to remove this comment after turning `getCommitInfo` into `evalCommits`

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/genqlient.graphql:
```diff
@@ -36,6 +36,41 @@ query checkLogin($login: String!) {
+          ... on Commit {
+            id
+            history(first: $commitFirst, after: $commitCursor) {
+              # @genqlient(typename: "CommitNode")
```
@crobert-1 this is the query we use that gets populated into generated Go code. Instead of uses `edges` for commits, I adjusted the query and shortened the name. We still do a type check because of the `... on Commit` syntax on the query.

Commit:
[chore] slight refactor and comment updates, addded additional tests

Commit:
[chore] limitations docs and cleanup

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -206,6 +210,92 @@ func (ghs *githubScraper) getPullRequests(
+	client graphql.Client,
+	repoName string,
+	branch BranchNode,
+) (int, int, int64, error) {
```
Set named returns and made it consistent in the locations for readability. `adds` and `dels` have been changed to `additions` and `deletions` respectfully.

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -206,6 +210,92 @@ func (ghs *githubScraper) getPullRequests(
+		}
+
+		if len(c.Edges) == 0 {
+			break
```
I've updated this to be nodes by adjusting the query and added comments for clarity. `break` was intended.

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -206,6 +210,92 @@ func (ghs *githubScraper) getPullRequests(
+		}
+		cc = &c.PageInfo.EndCursor
+		if nPage == comPages {
+			e := c.GetEdges()
```
This has since changed, however, yes, there's some intricacies on the graphql query -> generated go code side. It's been a while since this code was originally written so I had to go back and re-eval some context. 😅

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -206,6 +210,92 @@ func (ghs *githubScraper) getPullRequests(
+			age = int64(time.Since(oldest).Seconds())
+		}
+		for b := 0; b < len(c.Edges); b++ {
+			adds = add(adds, c.Edges[b].Node.Additions)
```
You're correct. Updated!

adrielp commented:
@crobert-1 - ready for another pass. Appreciate the review. For some reason the diff on the review tab is cached for me from GitHub. I resolved a bunch of conversations so if you see something resolved that I said changed, but doesn't appear changed, let me know / clear your cache maybe. IDK why GitHub's not reflecting the changes on the review right now.

Commit:
Merge branch 'main' into branch-metrics

crobert-1 commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -190,22 +199,128 @@ func (ghs *githubScraper) getPullRequests(
+
+		// GraphQL could return empty commit nodes so here we confirm that
+		// commits were returned to prevent an index out of range error. This
+		// technically should  never be triggered because of other preceding
```
```suggestion
		// technically should never be triggered because of other preceding
```

crobert-1 commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -190,22 +199,128 @@ func (ghs *githubScraper) getPullRequests(
+	return int(numPages)
+}
+
+func add[T ~int | ~float64](a, b T) T {
```
This is no longer used from what I can tell.

crobert-1 commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -42,21 +42,56 @@ metrics:
+    attributes: [repository.name, branch.name]
+  git.repository.branch.commit.behindby_count:
+    enabled: true
+    description: Number of commits a branch is behing the default branch.
```
```suggestion
    description: Number of commits a branch is behind the default branch.
```

crobert-1 commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -42,21 +42,56 @@ metrics:
+    attributes: [repository.name, branch.name]
+  git.repository.branch.line.addition_count:
+    enabled: true
+    description: Count of lines added to code in a branch
```
Should we also add in the description of the `git.repository.branch.line` that these values are relative to the default branch?

adrielp replied:
Yea, I like that idea! I'll adjust the decsription.

crobert-1 commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -42,21 +42,56 @@ metrics:
+    gauge:
+      value_type: int
+    attributes: [repository.name, branch.name]
+  git.repository.branch.commit.aheadby_count:
```
I'm wondering if the metric name here should be `aheadby.count`. I'm not super well educated on semantic conventions for naming, but it seems like those options would fit better than the [current naming schema ruels](https://opentelemetry.io/docs/specs/semconv/general/metrics/#use-count-instead-of-pluralization-for-updowncounters).

(Same comment for the rest of the `_count` metric names)

adrielp replied:
Yup, agreed, I'll make that adjustment. That's actually what we already have in the source, I made the change locally on this PR because of a failed memory from another conversation. Thanks for sending the reference doc.

crobert-1 reviewed (commented):
Looks good, mostly just some nits.

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -108,12 +108,62 @@ func (ghs *githubScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+			// metrics
+			for _, branch := range branches {
+
+				// For each branch we will check to make sure it's not the
```
Both

adrielp committed:
[chore] update comments, add docs, remove uncalled code

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -190,22 +199,128 @@ func (ghs *githubScraper) getPullRequests(
+	var cursor *string
+	items := defaultReturnItems
+
+	// We're using BehindBy here because we're comparing against the target
```
added a new readme inside of the github scraper folder that's linked from the top level readme & from inside the code.

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -190,22 +199,118 @@ func (ghs *githubScraper) getPullRequests(
+			// getCommitData function is accurate.
+			items = branch.Compare.BehindBy % 100
+			if items == 0 {
+				items = 100
```
```suggestion
				items = defaultReturnItems
```

Commit:
[chore] Update receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go

Commit:
[chore] Apply suggestions from code review

Co-authored-by: Curtis Robert <crobert@splunk.com>

adrielp committed:
[chore] update changelog to reflect the updated metric names

adrielp committed:
Merge branch 'main' into branch-metrics

adrielp committed:
[chore] re-tidy after merge

adrielp commented:
The failed links check is complaining about me linking to the file in `main` which hasn't been merged yet, so doesn't exist. Would it be better to link to the commit hash instead of main or just merge in as the check will pass next go-around?

crobert-1 commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/README.md:
```diff
@@ -0,0 +1,119 @@
+# GitHub Limitations
```
Maybe add a link to this README in the top level README, where you're referencing the GitHub scraper?

adrielp commented on receiver/gitproviderreceiver/go.mod:
```diff
@@ -6,6 +6,8 @@ require (
 	github.com/google/go-cmp v0.6.0
 	github.com/google/go-github/v61 v61.0.0
+	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/golden v0.100.0
+	github.com/open-telemetry/opentelemetry-collector-contrib/pkg/pdatatest v0.99.0
```
Yea, that's odd. I would've expected the tidy up I had ran to catch it & the workflow too. 🤔 - will fix.

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -190,22 +199,118 @@ func (ghs *githubScraper) getPullRequests(
+			// when it's a completely full page, so if the remainder is 0 we'll
+			// reset to 100 to ensure the items request sent to the
+			// getCommitData function is accurate.
+			items = branch.Compare.BehindBy % 100
```
Yup, agreed, I like it!

adrielp commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -42,21 +42,56 @@ metrics:
+    attributes: [repository.name, branch.name]
+  git.repository.branch.line.addition.count:
+    enabled: true
+    description: Count of lines added to code in a branch relative to the default branch (trunk).
```
I like number as well, will adjust!

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/README.md:
```diff
@@ -0,0 +1,119 @@
+  possible.
+* Both APIs have primary and secondary rate limits applied to them. The default
+  rate limit for GraphQL API is 5,000 points per hour when authenticated with a
+  user personal access toke (PAT). If using the [GitHub App Auth
```
I think i'm going to change to `... authenticated with a GitHub personal access token (PAT).` Instead.

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/README.md:
````diff
@@ -0,0 +1,119 @@
+  collection interval in seconds is:
+
+```math
+\text{collection\_interval (seconds)} = \frac{4n}{r/3600} + 300
````
It was a buffer. I've removed it.

adrielp committed:
[chore] adjust comments, docs, and descriptions based on feedback

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/README.md:
````diff
@@ -0,0 +1,119 @@
+\end{aligned}
+```
+
+$r$ is likely 5000 but there are factors that can change this,
````
Removed & cleaned up. let me know if it all makes sense now.

adrielp committed:
Merge branch 'main' into branch-metrics

crobert-1 commented on receiver/gitproviderreceiver/README.md:
```diff
@@ -19,35 +19,52 @@ The current default set of metrics common across all vendors can be found in [do
 These default metrics can be used as leading indicators to the DORA metrics; helping provide insight into modern-day engineering practices.
 
-## GitHub Metrics
+## GitHub Scraper
```
Nit: It makes more sense to me to have `Getting Started` first, then the `GitHub Scraper` section below the list of scrapers. Personal preference though, your call.

adrielp replied:
Yea, that makes sense!

adrielp replied:
Swapped it around.

crobert-1 commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -41,48 +41,83 @@ attributes:
+    attributes: [repository.name, branch.name]
+  git.repository.branch.line.addition.count:
+    enabled: true
+    description: The number of lines added to code in a branch relative to the default branch (trunk).
```
Nit: This includes doc changes, dep files, etc, right? (Like lines in a PR?) I'm wondering if there's a more generic way to word it than "code" in these metrics. I'll leave it up to you though, I assume it's probably clear enough.

adrielp replied:
I'm not sure. GitHub's GrapQL describes it as:

> The number of additions in this commit.

Maybe we should change it to:

```suggestion
    description: The number of lines added in a branch relative to the default branch (trunk).
```

adrielp replied:
Updated the description for both lines added and lines deleted.

crobert-1 reviewed (approved):
Just a couple nits, thanks for all of your work here!

Commit:
[chore] update comment and readme based on feedback

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -56,21 +60,26 @@ func (ghs *githubScraper) getBranches(
 	for next := true; next; {
+		// We decided to not use the defaultReturnItems here because the number
+		// we've the GitHub GraphQL internal server kill the connection on
+		// larger queries for branch data around the 80 per page mark.
```
Just pushed an update, hopefully it makes clearer sense now.

adrielp committed:
[chore] slight metric description re-wording

adrielp committed:
Merge branch 'main' into branch-metrics

adrielp committed:
[chore] resolve conflicts and re-tidy

adrielp commented:
> [<img alt="CLA Missing ID" width="328" height="28" src="https://camo.githubusercontent.com/83fc0261dbb20a7b16cc05d156131773ab142311e3e272338a23d3a26129bdc3/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6d697373696e672d69642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/196414933/32812/#/?version=2) [<img alt="CLA Not Signed" width="328" height="28" src="https://camo.githubusercontent.com/031843f73aecaf0b300b9aef2fc62377af0790eac726b513bf4ab20b714fbf8a/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6e6f742d7369676e65642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/196414933/32812/#/?version=2)
> 
> * ✅ login: adrielp / name: Adriel Perkins  ([c82c97b](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/c82c97b87e7f9368e5205724c0ab9009fa4f9ddf), [f2db263](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/f2db26388310f25d7dbcf2e6e2ecfe73e4da594e), [8c24669](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/8c2466932a3a6505a61d5b3d327f4e112442c8c1), [a312f79](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/a312f7928f939338fb0da254ca044e34728f9be7), [eb6ab5e](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/eb6ab5ecd5374717e0450b2c84748c5ac1e91215), [dbe9865](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/dbe98657431b058d1a6c87b3e3cc1bad4c5099dd), [a4cce65](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/a4cce659c7fc1bd08453a70ba2b694ebacef38b3), [ae25cdb](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/ae25cdbac72223066f523a4809430544a4d982fc), [b906a06](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/b906a06f590c69f79e4fb7994bce0d13d4f3b44f), [d8bf63d](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/d8bf63d2ac58d1f2e2e7856a94d6bf980ded8b06), [04aee28](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/04aee2851832995c738fcc25fc103562bbe26e26), [f0370b5](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/f0370b51fc6f2cf4ba930e8ce9541ade24682a66), [64f3884](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/64f388473677ba5301b06c4c4b92a78ca3f9b3fe), [b4b579b](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/b4b579b6a5c57e181a34b0c1e9dffe6f2a078f5a), [23fae24](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/23fae24a72e953dbc4372e17b9bdb40dd1d0e23c), [263971b](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/263971b2939dc4b9d742c4e7e81ab540b4f3beae), [5db15e3](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/5db15e36eca78b4b2fa39d75b3564876e10276ae), [b7317df](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/b7317df5f9ed64fb2f7a7a453538fe8515f3356f), [753257a](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/753257a2d0302982db34a7a5798e6b001b8cebb4), [6211eea](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/6211eea3f644741fed40d70bf499d5632d565214), [3c5afdc](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/3c5afdc7c3dfcdca7ea0598e9cd17b178d07c964), [194c68c](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/194c68c83138c236e056adc449621f3b28f3a4a9), [1084ac3](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/1084ac3bb7622b7b441a43a1f6a885f13e3645ac), [415a590](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/415a590479c550e6b3ea293e23aaf5233b4d7b00), [cf1f39d](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/cf1f39d32377a96c631a72fbace408dbe0e6a12c))
> * ❌ The email address for the commit ([1084ac3](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/1084ac3bb7622b7b441a43a1f6a885f13e3645ac)) is not linked to the GitHub account, preventing the EasyCLA check. Consult [this Help Article](https://confluence.linuxfoundation.org/pages/viewpage.action?pageId=86641302) and [GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve. (To view the commit's email address, add .patch at the end of this PR page's URL.) For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

@crobert-1  - looks like this is actually complaining about the co-author commit from the suggestion. I take it there's no chance to sign the CLA with your splunk email? That commit is a ways back.

adrielp commented:
> [<img alt="CLA Missing ID" width="328" height="28" src="https://camo.githubusercontent.com/83fc0261dbb20a7b16cc05d156131773ab142311e3e272338a23d3a26129bdc3/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6d697373696e672d69642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/196414933/32812/#/?version=2) [<img alt="CLA Not Signed" width="328" height="28" src="https://camo.githubusercontent.com/031843f73aecaf0b300b9aef2fc62377af0790eac726b513bf4ab20b714fbf8a/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6e6f742d7369676e65642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/196414933/32812/#/?version=2)
> 
> * ✅ login: adrielp / name: Adriel Perkins  ([c82c97b](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/c82c97b87e7f9368e5205724c0ab9009fa4f9ddf), [f2db263](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/f2db26388310f25d7dbcf2e6e2ecfe73e4da594e), [8c24669](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/8c2466932a3a6505a61d5b3d327f4e112442c8c1), [a312f79](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/a312f7928f939338fb0da254ca044e34728f9be7), [eb6ab5e](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/eb6ab5ecd5374717e0450b2c84748c5ac1e91215), [dbe9865](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/dbe98657431b058d1a6c87b3e3cc1bad4c5099dd), [a4cce65](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/a4cce659c7fc1bd08453a70ba2b694ebacef38b3), [ae25cdb](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/ae25cdbac72223066f523a4809430544a4d982fc), [b906a06](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/b906a06f590c69f79e4fb7994bce0d13d4f3b44f), [d8bf63d](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/d8bf63d2ac58d1f2e2e7856a94d6bf980ded8b06), [04aee28](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/04aee2851832995c738fcc25fc103562bbe26e26), [f0370b5](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/f0370b51fc6f2cf4ba930e8ce9541ade24682a66), [64f3884](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/64f388473677ba5301b06c4c4b92a78ca3f9b3fe), [b4b579b](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/b4b579b6a5c57e181a34b0c1e9dffe6f2a078f5a), [23fae24](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/23fae24a72e953dbc4372e17b9bdb40dd1d0e23c), [263971b](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/263971b2939dc4b9d742c4e7e81ab540b4f3beae), [5db15e3](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/5db15e36eca78b4b2fa39d75b3564876e10276ae), [b7317df](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/b7317df5f9ed64fb2f7a7a453538fe8515f3356f), [753257a](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/753257a2d0302982db34a7a5798e6b001b8cebb4), [6211eea](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/6211eea3f644741fed40d70bf499d5632d565214), [3c5afdc](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/3c5afdc7c3dfcdca7ea0598e9cd17b178d07c964), [194c68c](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/194c68c83138c236e056adc449621f3b28f3a4a9), [1084ac3](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/1084ac3bb7622b7b441a43a1f6a885f13e3645ac), [415a590](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/415a590479c550e6b3ea293e23aaf5233b4d7b00), [cf1f39d](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/cf1f39d32377a96c631a72fbace408dbe0e6a12c))
> * ❌ The email address for the commit ([1084ac3](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/1084ac3bb7622b7b441a43a1f6a885f13e3645ac)) is not linked to the GitHub account, preventing the EasyCLA check. Consult [this Help Article](https://confluence.linuxfoundation.org/pages/viewpage.action?pageId=86641302) and [GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve. (To view the commit's email address, add .patch at the end of this PR page's URL.) For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).

@crobert-1  - not sure why this one is showing up. 

I checked that commit (it's a ways back) and it has my name and email which is signed (other commits prove that). I thought at first it was your email in the co-authored section, but your name & email on that match other commits in the main branch that have passed fine.

crobert-1 commented:
> > [<img alt="CLA Missing ID" width="328" height="28" src="https://camo.githubusercontent.com/83fc0261dbb20a7b16cc05d156131773ab142311e3e272338a23d3a26129bdc3/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6d697373696e672d69642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/196414933/32812/#/?version=2) [<img alt="CLA Not Signed" width="328" height="28" src="https://camo.githubusercontent.com/031843f73aecaf0b300b9aef2fc62377af0790eac726b513bf4ab20b714fbf8a/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f636c612d70726f6a6563742d6c6f676f2d70726f642f636c612d6e6f742d7369676e65642e737667">](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/9522512/196414933/32812/#/?version=2)
> > 
> > * ✅ login: adrielp / name: Adriel Perkins  ([c82c97b](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/c82c97b87e7f9368e5205724c0ab9009fa4f9ddf), [f2db263](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/f2db26388310f25d7dbcf2e6e2ecfe73e4da594e), [8c24669](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/8c2466932a3a6505a61d5b3d327f4e112442c8c1), [a312f79](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/a312f7928f939338fb0da254ca044e34728f9be7), [eb6ab5e](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/eb6ab5ecd5374717e0450b2c84748c5ac1e91215), [dbe9865](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/dbe98657431b058d1a6c87b3e3cc1bad4c5099dd), [a4cce65](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/a4cce659c7fc1bd08453a70ba2b694ebacef38b3), [ae25cdb](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/ae25cdbac72223066f523a4809430544a4d982fc), [b906a06](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/b906a06f590c69f79e4fb7994bce0d13d4f3b44f), [d8bf63d](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/d8bf63d2ac58d1f2e2e7856a94d6bf980ded8b06), [04aee28](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/04aee2851832995c738fcc25fc103562bbe26e26), [f0370b5](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/f0370b51fc6f2cf4ba930e8ce9541ade24682a66), [64f3884](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/64f388473677ba5301b06c4c4b92a78ca3f9b3fe), [b4b579b](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/b4b579b6a5c57e181a34b0c1e9dffe6f2a078f5a), [23fae24](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/23fae24a72e953dbc4372e17b9bdb40dd1d0e23c), [263971b](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/263971b2939dc4b9d742c4e7e81ab540b4f3beae), [5db15e3](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/5db15e36eca78b4b2fa39d75b3564876e10276ae), [b7317df](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/b7317df5f9ed64fb2f7a7a453538fe8515f3356f), [753257a](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/753257a2d0302982db34a7a5798e6b001b8cebb4), [6211eea](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/6211eea3f644741fed40d70bf499d5632d565214), [3c5afdc](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/3c5afdc7c3dfcdca7ea0598e9cd17b178d07c964), [194c68c](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/194c68c83138c236e056adc449621f3b28f3a4a9), [1084ac3](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/1084ac3bb7622b7b441a43a1f6a885f13e3645ac), [415a590](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/415a590479c550e6b3ea293e23aaf5233b4d7b00), [cf1f39d](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/cf1f39d32377a96c631a72fbace408dbe0e6a12c))
> > * ❌ The email address for the commit ([1084ac3](https://github.com/open-telemetry/opentelemetry-collector-contrib/commit/1084ac3bb7622b7b441a43a1f6a885f13e3645ac)) is not linked to the GitHub account, preventing the EasyCLA check. Consult [this Help Article](https://confluence.linuxfoundation.org/pages/viewpage.action?pageId=86641302) and [GitHub Help](https://help.github.com/en/github/committing-changes-to-your-project/why-are-my-commits-linked-to-the-wrong-user) to resolve. (To view the commit's email address, add .patch at the end of this PR page's URL.) For further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).
> 
> @crobert-1 - not sure why this one is showing up.
> 
> I checked that commit (it's a ways back) and it has my name and email which is signed (other commits prove that). I thought at first it was your email in the co-authored section, but your name & email on that match other commits in the main branch that have passed fine.
> 
> EDIT: must be old/cached (even though it says updated an hour ago) because the check that runs in the workflows is passing.

It's an EasyCLA bug, unrelated to this PR. (This was hit on another PR as well, here's [the comment with more context](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32981#issuecomment-2120923144).)

adrielp committed:
[chore] update logging statements and add mutex lock

Mutex lock prevents metadata builder bug in SetStartTimestamp
when recording metrics within a goroutine.

adrielp committed:
Merge branch 'main' into branch-metrics

adrielp committed:
[chore] rerun tidy's after merge conflict resolution

adrielp commented:
@crobert-1 - made a couple small logging statement updates & fixed a bug with metric build record statements running from within goroutines (mutex lock) - just wanted to let you know the change while waiting for another review.

andrzej-stencel commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -41,48 +41,83 @@ attributes:
+    gauge:
+      value_type: int
+    attributes: [repository.name, branch.name]
+  git.repository.branch.commit.behindby.count:
```
```suggestion
  git.repository.branch.commits_behind:
```

andrzej-stencel commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -41,48 +41,83 @@ attributes:
     gauge:
       value_type: int
     attributes: [repository.name]
+  git.repository.branch.time:
```
Naming this `git.repository.branch.time_open` would make it more in line with `git.repository.pull_request.time_open`, but I'm not sure if it's better. What do you think?

adrielp replied:
> time - an instrument that measures passage of time should be called entity.time. For example, system.cpu.time with attribute state = idle | user | system | .... time measurements are not necessarily wall time and can be less than or greater than the real wall time between measurements.
> 
> time instruments are a special case of usage metrics, where the limit can usually be calculated as the sum of time over all attribute values. utilization for time instruments can be derived automatically using metric event timestamps. For example, system.cpu.utilization is defined as the difference in system.cpu.time measurements divided by the elapsed time and number of CPUs.

That's from the metric doc on the convention. That's why I went with `time` originally. If we feel `time_open` is more clear & inline with the convention then I'm good to change this to `time_open`. 

Btw, every time I change these I go in and change the underlying dashboards that are published on the grafana dashboard market to support these new metric names. And I plan on going through and changing this all again to match the `scm...` attribute convention once that gets merged.

andrzej-stencel commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -98,33 +98,72 @@ func (ghs *githubScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+				// for more information as to why we do not emit metrics for
+				// the default branch (trunk) nor any branch with no changes to
+				// it.
+				if branch.Name == branch.Repository.DefaultBranchRef.Name || branch.Compare.BehindBy == 0 {
```
Why do you use the `BehindBy` and not `AheadBy` to find branches that have no changes? I'd think that if a branch has no changes, its `AheadBy` is equal to zero, not necessarily the `BehindBy`. Am I misunderstanding something?

andrzej-stencel replied:
Oh I see below is a reference to the GitHub API limitations doc. I read it and I still don't understand why `BehindBy` and `AheadBy` are swapped, but I'll take your word for it. :sweat_smile:

adrielp replied:
Branches get compared against trunk, and therefor GitHub responds with `trunk is behind your branch by` which is the same as saying `your branch is ahead trunk by`.

adrielp commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -41,48 +41,83 @@ attributes:
+    gauge:
+      value_type: int
+    attributes: [repository.name, branch.name]
+  git.repository.branch.line.deletion.count:
```
I could get behind `...lines{state=deleted}` . I think `line(s)` is more in-line with an attribute/characteristic than a namespace for now, though I disagree with that on the `commit` side.

adrielp committed:
Merge branch 'main' into branch-metrics

adrielp committed:
[chore] run checks and tidy's after resolving conflicts

andrzej-stencel commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -41,48 +41,83 @@ attributes:
+    gauge:
+      value_type: int
+    attributes: [repository.name, branch.name]
+  git.repository.branch.commit.aheadby.count:
```
OK, given that the component is in development status, let's merge this as is. We can perform renames later if required.

adrielp committed:
Merge branch 'main' into branch-metrics

adrielp committed:
[chore] update go.mod and go.sum files

crobert-1 commented:
_Only failing action is check-links, but the links it's failing on are being added in this PR, so I think it's acceptable to merge as-is._

andrzej-stencel commented:
> _Only failing action is check-links, but the links it's failing on are being added in this PR, so I think it's acceptable to merge as-is._

The link would be valid if it was a relative link:

```markdown
[ghsread]: ./internal/scraper/githubscraper/README.md#github-limitations
```

But that's not blocking.
